src/Microsoft.AspNet.Mvc.Core/Routing/KnownRouteValueConstraint.cs

```diff
@@ -43,20 +42,20 @@ public class KnownRouteValueConstraint : IRouteConstraint
                 throw new ArgumentNullException(nameof(values));
             }
 
-            object value;
-            if (values.TryGetValue(routeKey, out value))
+            object obj;
+            if (values.TryGetValue(routeKey, out obj))
             {
-                var valueAsString = value as string;
-
-                if (valueAsString != null)
+                var value = obj as string;
+                if (value != null)
                 {
                     var allValues = GetAndCacheAllMatchingValues(routeKey, httpContext);
-                    var match = allValues.Any(existingRouteValue =>
-                                                existingRouteValue.Equals(
-                                                                    valueAsString,
-                                                                    StringComparison.OrdinalIgnoreCase));
-
-                    return match;
+                    foreach (var existingValue in allValues)
+                    {
+                        if (string.Equals(value, existingValue, StringComparison.OrdinalIgnoreCase))
+                        {
+                            return true;
+                        }
+                    }
```

Comment:
This is linq on a pretty hot path. We haven't been testing with this constraint, but it's definitely worth fixing.